File: src/spiral/managers/project.py
from __future__ import annotations

import re
import shutil
import subprocess
from pathlib import Path
from typing import Any

from spiral.config import Config


class ProjectManager:
    def __init__(self, config: Config):
        self.config = config
        self.project_dir = config.PROJECT_DIR

    def read_adr(self) -> str:
        return self.config.ADR_PATH.read_text()

    def read_agents_md(self) -> str:
        return self.config.AGENTS_PATH.read_text()

    def update_adr(self, section: str, status: str) -> None:
        adr = self.config.ADR_PATH.read_text()
        marker = f"## {section}"
        if marker not in adr:
            return
        lines = adr.split("\n")
        new_lines = []
        found = False
        status_updated = False
        for i, line in enumerate(lines):
            if line.strip().startswith(marker):
                found = True
            if found and line.strip().startswith("**Status:**"):
                line = f"**Status:** {status}"
                found = False
                status_updated = True
            new_lines.append(line)
        if not status_updated:
            for i, line in enumerate(new_lines):
                if line.strip().startswith(marker):
                    new_lines.insert(i + 1, f"**Status:** {status}")
                    break
        self.config.ADR_PATH.write_text("\n".join(new_lines))

    def mark_adr_done(self, section: str) -> None:
        self.update_adr(section, "Done")

    def append_adr_section(self, title: str, description: str) -> None:
        adr = self.config.ADR_PATH.read_text()
        new_section = f"\n\n## {title}\n\n{description}\n\n**Status:** Pending\n"
        self.config.ADR_PATH.write_text(adr + new_section)

    def get_adr_section_names(self) -> list[str]:
        adr = self.config.ADR_PATH.read_text()
        sections = []
        for line in adr.split("\n"):
            stripped = line.strip()
            if stripped.startswith("## "):
                sections.append(stripped[3:].strip())
        return sections

    def file_exists(self, path: str) -> bool:
        return (self.project_dir / path).exists()

    def read_file(
        self, path: str, offset: int | None = None, limit: int | None = None
    ) -> str:
        f = self.project_dir / path
        if not f.exists():
            return ""
        if offset is None and limit is None:
            return f.read_text()
        lines = f.read_text().splitlines()
        start = max(0, (offset or 1) - 1)
        end = start + limit if limit else len(lines)
        selected = lines[start:end]
        numbered = [f"{start + i + 1}: {line}" for i, line in enumerate(selected)]
        return "\n".join(numbered)

    def write_file(self, path: str, content: str) -> None:
        f = self.project_dir / path
        f.parent.mkdir(parents=True, exist_ok=True)
        f.write_text(content)

    def edit_file(
        self, path: str, old_string: str, new_string: str, replace_all: bool = False
    ) -> str:
        f = self.project_dir / path
        if not f.exists():
            raise FileNotFoundError(f"File not found: {path}")
        content = f.read_text()
        count = content.count(old_string)
        if count == 0:
            raise ValueError(f"old_string not found in {path}")
        if count > 1 and not replace_all:
            raise ValueError(
                f"old_string found {count} times in {path}; "
                "use replaceAll=true or provide more context"
            )
        if replace_all:
            new_content = content.replace(old_string, new_string)
        else:
            new_content = content.replace(old_string, new_string, 1)
        f.write_text(new_content)
        replaced = count if replace_all else 1
        return f"Replaced {replaced} occurrence(s) in {path}"

    def glob(self, pattern: str, path: str | None = None) -> list[str]:
        base = self.project_dir / (path or "")
        return sorted(str(p.relative_to(self.project_dir)) for p in base.glob(pattern))

    def grep(
        self,
        pattern: str,
        include: str = "*",
        path: str | None = None,
        max_results: int = 100,
    ) -> list[dict[str, Any]]:
        base = self.project_dir / (path or "")
        if not base.exists():
            return []
        rg = shutil.which("rg")
        if rg:
            cmd = [
                rg,
                "--line-number",
                "--no-heading",
                "--color=never",
            ]
            if include != "*":
                cmd.extend(["--glob", include])
            cmd.extend([pattern, str(base)])
            result = subprocess.run(
                cmd, capture_output=True, text=True, timeout=30, check=False
            )
            matches: list[dict[str, Any]] = []
            for line in result.stdout.splitlines()[:max_results]:
                parts = line.split(":", 2)
                if len(parts) >= 3:
                    file_path, line_no, content = parts[0], parts[1], parts[2]
                    rel = str(Path(file_path).relative_to(self.project_dir))
                    matches.append(
                        {"file": rel, "line": int(line_no), "content": content}
                    )
            return matches
        regex = re.compile(pattern)
        results: list[dict[str, Any]] = []
        for fp in base.rglob("*"):
            if not fp.is_file():
                continue
            if include != "*" and not fp.match(include):
                continue
            try:
                for i, line in enumerate(fp.read_text().splitlines(), 1):
                    if regex.search(line):
                        rel = str(fp.relative_to(self.project_dir))
                        results.append({"file": rel, "line": i, "content": line})
                        if len(results) >= max_results:
                            return results
            except (OSError, UnicodeDecodeError):
                continue
        return results

    def run_command(
        self, cmd: list[str], cwd: str | None = None
    ) -> subprocess.CompletedProcess:
        return subprocess.run(
            cmd,
            cwd=cwd or str(self.project_dir),
            capture_output=True,
            text=True,
            timeout=120,
            check=False,
        )

    def list_project_files(self, pattern: str = "**/*.py") -> list[str]:
        return [
            str(p.relative_to(self.project_dir)) for p in self.project_dir.glob(pattern)
        ]

    def get_project_context(self) -> dict[str, Any]:
        return {
            "adr": self.read_adr(),
            "agents_md": self.read_agents_md(),
            "files": self.list_project_files(),
            "project_dir": str(self.project_dir),
        }

File: src/managers/project.ts
import { spawn } from "node:child_process";
import { readFile, writeFile, mkdir, access } from "node:fs/promises";
import { constants } from "node:fs";
import path from "node:path";
import { Config } from "../config.js";
import type { GrepMatch } from "../types.js";

async function fileExists(p: string): Promise<boolean> {
  try {
    await access(p, constants.F_OK);
    return true;
  } catch {
    return false;
  }
}

async function safeReadFile(p: string): Promise<string> {
  try {
    return await readFile(p, "utf-8");
  } catch {
    return "";
  }
}

export class ProjectManager {
  config: Config;
  projectDir: string;

  constructor(config: Config) {
    this.config = config;
    this.projectDir = config.projectDir;
  }

  async readAdr(): Promise<string> {
    return readFile(this.config.adrPath, "utf-8");
  }

  async readAgentsMd(): Promise<string> {
    return readFile(this.config.agentsPath, "utf-8");
  }

  async updateAdr(section: string, status: string): Promise<void> {
    const adr = await this.readAdr();
    const marker = `## ${section}`;
    if (!adr.includes(marker)) return;
    const lines = adr.split("\n");
    const newLines: string[] = [];
    let found = false;
    let statusUpdated = false;
    for (const line of lines) {
      if (line.trim().startsWith(marker)) found = true;
      if (found && line.trim().startsWith("**Status:**")) {
        newLines.push(`**Status:** ${status}`);
        found = false;
        statusUpdated = true;
      } else {
        newLines.push(line);
      }
    }
    if (!statusUpdated) {
      for (let i = 0; i < newLines.length; i++) {
        if (newLines[i].trim().startsWith(marker)) {
          newLines.splice(i + 1, 0, `**Status:** ${status}`);
          break;
        }
      }
    }
    await writeFile(this.config.adrPath, newLines.join("\n"), "utf-8");
  }

  async markAdrDone(section: string): Promise<void> {
    await this.updateAdr(section, "Done");
  }

  async appendAdrSection(title: string, description: string): Promise<void> {
    const adr = await this.readAdr();
    const newSection = `\n\n## ${title}\n\n${description}\n\n**Status:** Pending\n`;
    await writeFile(this.config.adrPath, adr + newSection, "utf-8");
  }

  async getAdrSectionNames(): Promise<string[]> {
    const adr = await this.readAdr();
    const sections: string[] = [];
    for (const line of adr.split("\n")) {
      const stripped = line.trim();
      if (stripped.startsWith("## ")) sections.push(stripped.slice(3).trim());
    }
    return sections;
  }

  async fileExists(relPath: string): Promise<boolean> {
    return fileExists(path.join(this.projectDir, relPath));
  }

  async readFile(relPath: string, offset?: number, limit?: number): Promise<string> {
    const f = path.join(this.projectDir, relPath);
    if (!(await fileExists(f))) return "";
    if (offset === undefined && limit === undefined) {
      return safeReadFile(f);
    }
    const content = await safeReadFile(f);
    const lines = content.split("\n");
    const start = Math.max(0, (offset ?? 1) - 1);
    const end = limit ? start + limit : lines.length;
    const selected = lines.slice(start, end);
    return selected.map((line, i) => `${start + i + 1}: ${line}`).join("\n");
  }

  async writeFile(relPath: string, content: string): Promise<void> {
    const f = path.join(this.projectDir, relPath);
    await mkdir(path.dirname(f), { recursive: true });
    await writeFile(f, content, "utf-8");
  }

  async editFile(
    relPath: string,
    oldString: string,
    newString: string,
    replaceAll = false,
  ): Promise<string> {
    const f = path.join(this.projectDir, relPath);
    if (!(await fileExists(f))) throw new Error(`File not found: ${relPath}`);
    const content = await safeReadFile(f);
    const count = content.split(oldString).length - 1;
    if (count === 0) throw new Error(`old_string not found in ${relPath}`);
    if (count > 1 && !replaceAll)
      throw new Error(
        `old_string found ${count} times in ${relPath}; use replaceAll=true or provide more context`,
      );
    const newContent = replaceAll
      ? content.split(oldString).join(newString)
      : content.replace(oldString, newString);
    await writeFile(f, newContent, "utf-8");
    return `Replaced ${replaceAll ? count : 1} occurrence(s) in ${relPath}`;
  }

  async glob(pattern: string, subdir?: string): Promise<string[]> {
    const base = subdir ? path.join(this.projectDir, subdir) : this.projectDir;
    return globImpl(base, pattern, this.projectDir);
  }

  async grep(
    pattern: string,
    opts: { include?: string; subdir?: string; maxResults?: number } = {},
  ): Promise<GrepMatch[]> {
    const base = opts.subdir ? path.join(this.projectDir, opts.subdir) : this.projectDir;
    const include = opts.include ?? "*";
    const max = opts.maxResults ?? 100;
    return grepImpl(base, pattern, include, max, this.projectDir);
  }

  runCommand(
    cmd: string[],
    cwd?: string,
  ): Promise<{ stdout: string; stderr: string; code: number }> {
    return new Promise((resolve) => {
      const proc = spawn(cmd[0]!, cmd.slice(1), {
        cwd: cwd ?? this.projectDir,
        stdio: ["pipe", "pipe", "pipe"],
      });
      let stdout = "";
      let stderr = "";
      proc.stdout?.on("data", (d: Buffer) => (stdout += d.toString()));
      proc.stderr?.on("data", (d: Buffer) => (stderr += d.toString()));
      proc.on("close", (code: number | null) => resolve({ stdout, stderr, code: code ?? 0 }));
      proc.on("error", () => resolve({ stdout, stderr, code: 1 }));
    });
  }

  async listProjectFiles(pattern = "**/*.ts"): Promise<string[]> {
    return this.glob(pattern);
  }

  async getProjectContext(): Promise<Record<string, unknown>> {
    return {
      adr: await this.readAdr(),
      agents_md: await this.readAgentsMd(),
      files: await this.listProjectFiles(),
      project_dir: this.projectDir,
    };
  }
}

async function globImpl(base: string, pattern: string, projectDir: string): Promise<string[]> {
  const { glob } = await import("node:fs/promises");
  try {
    const results: string[] = [];
    for await (const entry of glob(pattern, { cwd: base, withFileTypes: false })) {
      results.push(path.relative(projectDir, path.join(base, entry)));
    }
    return results.sort();
  } catch {
    return [];
  }
}

async function grepImpl(
  base: string,
  pattern: string,
  include: string,
  max: number,
  projectDir: string,
): Promise<GrepMatch[]> {
  const { glob } = await import("node:fs/promises");
  const regex = new RegExp(pattern);
  const results: GrepMatch[] = [];
  try {
    const filePattern = include === "*" ? "**/*" : `**/${include}`;
    for await (const entry of glob(filePattern, { cwd: base, withFileTypes: false })) {
      if (results.length >= max) break;
      const fullPath = path.join(base, entry);
      try {
        const content = await readFile(fullPath, "utf-8");
        const lines = content.split("\n");
        for (let i = 0; i < lines.length; i++) {
          if (results.length >= max) break;
          if (regex.test(lines[i]!)) {
            results.push({
              file: path.relative(projectDir, fullPath),
              line: i + 1,
              content: lines[i]!,
            });
          }
        }
      } catch {
        continue;
      }
    }
  } catch {
    return [];
  }
  return results;
}
